x-pack/elastic-agent/pkg/agent/control/server/server.go

```diff
@@ -210,70 +213,233 @@ func (s *Server) ProcMeta(ctx context.Context, _ *proto.Empty) (*proto.ProcMetaR
-			}
 			endpoint := monitoring.MonitoringEndpoint(spec, runtime.GOOS, rk)
-			if strings.HasPrefix(endpoint, "unix://") {
-				client.Transport = &http.Transport{
-					Proxy:       nil,
-					DialContext: socket.DialContext(strings.TrimPrefix(endpoint, "unix://")),
-				}
-				endpoint = "unix"
-			} else if strings.HasPrefix(endpoint, "npipe://") {
-				client.Transport = &http.Transport{
-					Proxy:       nil,
-					DialContext: socket.DialContext(strings.TrimPrefix(endpoint, "npipe:///")),
-				}
-				endpoint = "npipe"
-			}
+			client := newSocketRequester(n, rk, endpoint)
+
+			procMeta := client.procMeta(ctx)
+			resp.Procs = append(resp.Procs, procMeta)
+		}
+	}
+	return resp, nil
+}
 
-			res, err := client.Get("http://" + endpoint + "/")
-			if err != nil {
-				procMeta.Error = err.Error()
-				resp.Procs = append(resp.Procs, procMeta)
+// Pprof returns /debug/pprof data for the requested applicaiont-route_key or all running applications.
+func (s *Server) Pprof(ctx context.Context, req *proto.PprofRequest) (*proto.PprofResponse, error) {
+	if s.routeFn == nil {
+		return nil, errors.New("route function is nil")
+	}
+
+	dur, err := time.ParseDuration(req.TraceDuration)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse trace duration: %w", err)
+	}
+
+	resp := &proto.PprofResponse{
+		Results: []*proto.PprofResult{},
+	}
+
+	var wg sync.WaitGroup
+	ch := make(chan *proto.PprofResult, 1)
+
+	// retrieve elastic-agent pprof data if requested or application is unspecified.
+	if req.AppName == "" || req.AppName == "elastic-agent" {
+		endpoint := beats.AgentMonitoringEndpoint(runtime.GOOS, nil) //TODO get monitoring conf?
+		c := newSocketRequester("elastic-agent", "", endpoint)
+		for _, opt := range req.PprofType {
+			wg.Add(1)
+			go func(opt proto.PprofOption) {
+				res := c.getPprof(ctx, opt, dur)
+				ch <- res
+				wg.Done()
+			}(opt)
+		}
+	}
+
+	routes := s.routeFn()
+	for _, rk := range routes.Keys() {
+		// Skip the rk if it does not match and one has been requested.
+		if req.RouteKey != "" && req.RouteKey != rk {
+			continue
+		}
+		programs, ok := routes.Get(rk)
+		if !ok {
+			s.logger.With("route_key", rk).Warn("Unable to retrieve route.")
+			continue
+		}
+
+		sp, ok := programs.(specer)
+		if !ok {
+			s.logger.With("route_key", rk, "route", programs).Warn("Unable to cast route as specer.")
+			continue
+		}
+		specs := sp.Specs()
+		for n, spec := range specs {
+			// Skip the app if it does not match and one has been requested.
+			if req.AppName != "" && req.AppName != n {
 				continue
 			}
-			if res.StatusCode != 200 {
-				procMeta.Error = "response status is: " + res.Status
-				resp.Procs = append(resp.Procs, procMeta)
-				continue
+			endpoint := monitoring.MonitoringEndpoint(spec, runtime.GOOS, rk)
+			c := newSocketRequester(n, rk, endpoint)
+
+			// Launch a concurrent goroutine to gather all pprof endpoints from a socket.
+			for _, opt := range req.PprofType {
```

Comment:
We have here 3 for loops nested inside each other. Maybe we could extract some into methods which would make reviewing and potentially testing easier.

Reply:
i'll work on it